docs/style-guide.md

```diff
@@ -0,0 +1,1348 @@
+
+inline int pthread_spin_lock(pthread_spinlock_t *lock) {
+    OSSpinLockLock(lock);
+    return 0;
+}
+...
+#endif // __APPLE__
+```
+
+Similar principle applies when a component is completely missing.
+Use the API as if the component was there.
+Provide a header file with stubs for the missing functions.
+Write your code such that it works correctly if stubs are called
+instead of the missing component.
+PULSAR is written that way with respect to support for MPI and CUDA.
+It may not always be possible, but at least investigate the possibility.
+
+Naming
+======
+
+* Use abbreviations within reason. Follow common conventions.
+  Name a variable `retval` rather than `return_value`.
+
+* There are no special requirements for global variables,
+  which should be rare anyway, but if you use them, consider prefixing them
+  with `g_` or some other marker to easily distinguish it from local variables.
+
+File Names
+----------
+
+Most projects already have a convention.
+Follow the project's convention.
+
+In C++ the file name should match the class name.
+Consider using the Trilinos convention:
+`<NameSpace>_<ClassName>`, i.e., namespace using Pascal case,
+underscore, class name using Pascal case,
+e.g., `Magma_SomeClass.hh`, `Magma_SomeClass.cc`.
+
+Use `.h` and `.c` extensions for C files, and `.hh` and `.cc` for C++ files.
+
+Type Names
+----------
+
+* In C++ use Pascal case for class names, e.g., `MyExcitingClass`.
+
+* In C use camel case for type names followed by `_t`, e.g., `myExcitingType_t`
+  (NVIDIA's convention).
+
+Wrap enums and structs in typedefs and do not use `enum` and `struct` keywords
+when declaring variables.
+
+Variable Names
+--------------
+
+* The names of variables and data members are all lowercase,
+  with underscores between words.
+  Data members of classes (but not structs) additionally have trailing underscores.
+  For instance:
+
+    - `a_local_variable`,
+    - `a_struct_data_member`,
+    - `a_class_data_member_`.
+
+Constant Names
+--------------
+
+Use Pascal case for constants, i.e., `const int ArraySize`.
+
+> This is Microsoft's convention.
+> Google's convention is to use a leading `k`, i.e., `kArraySize`,
+> which seems a little arbitrary and looks somewhat awkward.
+
+Prefix global variables with `g_` and global constants with `G_`, e.g.,
+`g_scary_global_variable`, `G_ScaryGlobalConstant`.
+
+Function Names
+--------------
+
+Use snake case for C function names and C++ method names,
+e.g., `my_awesome_c_function`, `my_awesome_cpp_method`.
+
+> SLATE used lowerCamelCase for methods, but there has been discussion
+> to make them snake_case.
+
+Namespace Names
+---------------
+
+Namespace names are Pascal case (Trilinos, Microsoft).
+
+> SLATE uses lowercase for namespaces.
+
+Enumerator Names
+----------------
+
+Individual enumerators should be named like constants, i.e., Pascal case.
+
+Macro Names
+-----------
+
+1. Do not use macros.
+
+2. If you really need to use a macro, see 1.
+
```

Comment:
3. If you need to shorten a test code or a code for internal-use only then see rule no. 2. ;-)